client/state/test/utils.js

```diff
@@ -306,6 +306,14 @@ describe( 'utils', () => {
 			} );
 		} );
 
+		test( 'should handle keyNames referencing nested keys', () => {
+			const keyed = keyedReducer( 'person.name', age );
+			const action = { type: 'GROW', person: { name: 'Calypso' } };
+			expect( keyed( undefined, action ) ).to.eql( {
+				Calypso: 1,
+			} );
+		} );
```

Comment:
how about one also like `[ 'sites', 0, 'name']`

Reply:
currently disallowing them: https://github.com/Automattic/wp-calypso/pull/18805/files#diff-e76708e1e8c79e9a41cb0813e27eb468R81

I'm not sure it would make sense to readily support dynamically changing keynames either...

Reply:
cool. any reason for disallowing it?

Reply:
> I'm not sure it would make sense to readily support dynamically changing keyPaths either...
 
I think that the array notation encourages usage of variables.  It isn't more ergonomic for static keypaths but _is_ more clean for ones involving vars.  Sure people could interpolate  variables into strings, but I'd prefer to keep it this way for now.

I rather err on the side of smaller api